### PR TITLE
dash/dg2-column-class-name-templating

### PR DESCRIPTION
**Pull request by ddragula**
Added options allowing setting an additional class name with templating for cells.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1208156929604162

### CODE REVIEW

**Comment by stitot** [comment]
OK, have tested this. Some bugs, but won't comment for now, because:

Sorry for introducing complexity, but I believe we can simplify this and accomplish the same. Simpler in API, perhaps more complex behind the scenes:

Let's enable templating for `columns[].className` and not introduce new options. In the example below I check for values > 10 (red) and values < 10 (blue). The `td` receives one class name, while `th`  receives an aggregated list of all "hits" in that column. 

This way devs can style both `th` and `td` independently, and even `th.red.blue {background: purple}`. If they don't like the extra class names in the header they can override/remove these using `header`

```
<table>
    <thead>
        <tr>
            <th class="blue red">Column 1</th>
            <th class="blue">Column 2</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td class="blue">3</td>
            <td class="blue">7</td>
        </tr>
        <tr>
            <td class="red">15</td>
            <td class="blue">4</td>
        </tr>
        <tr>
            <td>10</td>
            <td>10</td>
        </tr>
        <tr>
            <td class="red">12</td>
            <td>10</td>
        </tr>
    </tbody>
</table>
```

**Comment by ddragula** [comment]
>Let's enable templating for columns[].className and not introduce new options.

I really want `columns.className[]` to be independent of templating for two reasons:
- it refers to the entire column, not to cells. It is also assigned to cells, but only because we do not have a separate HTML element that represents a column.
- it is assigned to the mocked element before the column is rendered (before loading the column and cells data), because it is used to calculate the column width.

**Summary:** I suggest leaving the new `cellClassName` option which handles templating and refers to individual cells in a column.

## Now about the headers

I can remove the `headerCellClassName` option and handle them with `cellClassName` as well, BUT even though `TableCell` and `HeaderCell` inherit from one abstract `Cell` class, they are two different classes with slightly different fields. And I assume that the way these options are used will be different.

I don't quite understand what you mean here:
> while th receives an aggregated list of all "hits" in that column

In the sense that each header cell gets one class name from each individual occurrence of the class name from cells belonging to a given column?

Currently, `headerCellClassName` takes as context a `HeaderCell` instance which contains a `cell.column.values` parameter that holds a reference to an array of all values ​​in a given column in the DataTable. If you want, you can iterate through them using `#each` in templating.

## Final thoughts
The `columns[].className` refers only to columns and should not be mixed with cells, And the other two new options refer to two different types of cells, so I wouldn't unify them.

If these three options were to remain, then alternatively I might suggest the following API structure:
```ts
columns: [{
  id: 'someColumn',
  className: 'custom-column-class-name',
  cells: {
    className: 'custom-cell-class-name-with-templating-{cell.value}',
    format: '',
    formatter: ''
  },
  header: {
    className: 'custom-header-cell-class-name-with-templating-{cell.column.values.length}',
    format: ''
  }
}]
```

instead of:

```ts
columns: [{
  id: 'someColumn',
  className: 'custom-column-class-name',
  cellClassName: 'custom-cell-class-name-with-templating-{cell.value}',
  cellFormat: '',
  cellFormatter: '',
  headerClassName: 'custom-header-cell-class-name-with-templating-{cell.column.values.length}',
  headerFormat: '',
}]
```


If you feel this adds too many options to the API then instead, adding values ​​on the fly can be done by hooking into the `setValues` method.

**Demo:** https://jsfiddle.net/BlackLabel/3dt9qemc/

**Comment by stitot** [comment]
Hmmmmm, I can't agree on the premise that `columns[].className` "refers to the entire column, not to cells". It is _perceived_ as though it refer to the entire column when there is no logic in `columns[].className`, but when logic is added it comes clear that it refer to the cell, not the entire column.

With no logic e.g `className: "red"` it always "hits" and adds classname to all cells, but we're still on cell level, not column. When adding logic, e.g. `className: "{#if (gt value 10)}red{/if}"` it only adds classname to some cells, and it's clearer that we are on cell level, not column.

If dev need both, meaning the _perceived_ column scope AND cell scope they could use e.g. `className: "red {#if (gt value 10)}redder{/if}"`

So I don't see the need for introducing more options here, but could be missing something!

Now, even though we provide the "extra service" of also adding these CSS classes to the column's `th` it's stricly a styling matter. They can choose to target them with CSS or not. If logic in `columns[].className` leads to several different class names in that column, by e.g. using `className: "{#if (gt value 10)}red{else}green{/if}"` then all these should be applied to the `th`, e.g. `<th class="red green"/>`.

Have to run, but let's discuss in person tomorrow if we are talking past each other.

**Comment by ddragula** [comment]
`className` without templating is important when setting the custom column width (it's assigned to the mock element).
https://github.com/highcharts/highcharts/blob/3698060fbb0b64fcb85bc71678d0fcea8df2bdc0/ts/DataGrid/Column.ts#L250-L252

The idea behind className was to be able to set width without having to use `td[data-cell-id="someCellId"]` in CSS. If we do formatting there, this functionality will have to be removed, and this will be the only way to set the column's width.

If we want to limit adding new options there, maybe it will be enough to just add a new event option `afterSetValue` like in the PoC: https://jsfiddle.net/BlackLabel/bdsve47j/ ?

**Comment by stitot** [comment]
OK, wasn't aware that className was used in the mock element. We'll continue this discussion offline.

File: docs/datagrid/understanding-datagrid.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ Header
 
 The [table header](https://api.highcharts.com/dashboards/#classes/DataGrid_DataGridTableHeader) is a special row, always on the top, containing the column IDs.
 Cells in the `header` are called `headerCell`. Their contents can be edited using the
-[`headerFormat` option](https://api.highcharts.com/dashboards/#interfaces/DataGrid_DataGridOptions.ColumnOptions#headerFormat).
+[`header.format` option](https://api.highcharts.com/dashboards/#interfaces/DataGrid_DataGridOptions.ColumnOptions#headerFormat).
```
Comment (sebastianbochan) [inline docs/datagrid/understanding-datagrid.md:38]: Is it valid url?

File: docs/datagrid/understanding-datagrid.md [hunk 2]
```diff
@@ -112,7 +116,7 @@ columnDefaults: {
 
 Click on a cell and change the value.
 
-If you declared `cellFormatter` or `cellFormat` props, it will be applied to the
+If you declared `celss.formatter` or `cells.format` props, it will be applied to the
```
Comment (sebastianbochan) [inline docs/datagrid/understanding-datagrid.md:119]: `celss.formatter` should be `cells.formatter`

File: ts/DataGrid/Options.ts [hunk 3]
```diff
@@ -181,34 +181,16 @@ export interface RowsSettings {
...
-     * @return
-     * A string to be concatenated in to the common cell's text.
+     * Options for all cells in the column.
      */
-    cellFormatter?: CellFormatterCallback;
+    cells?: ColumnCellOptions;
 
     /**
-     * The format of the column header. Use `{id}` to display the column id.
+     * Options for all the header cells in the column.
```
Comment (sebastianbochan) [inline ts/DataGrid/Options.ts:191]: `Options for the header cells in the columns` sounds better.

File: ts/DataGrid/Options.ts [hunk 4]
```diff
@@ -237,6 +219,74 @@ export interface ColumnOptions {
     resizing?: boolean;
 }
 
+/**
+ * Options for all cells in the column.
+ */
+export interface ColumnCellOptions {
+    /**
+     * Allows to define an additional class name to all table cells in the
```
Comment (sebastianbochan) [inline ts/DataGrid/Options.ts:227]: `Allows user to define`

File: ts/DataGrid/Options.ts [hunk 5]
```diff
@@ -237,6 +219,74 @@ export interface ColumnOptions {
...
+     * column. Applied only to cell that are in the table not the column header.
```
Comment (sebastianbochan) [inline ts/DataGrid/Options.ts:228]: `not in the column header`

File: ts/DataGrid/Querying/QueryingController.ts [hunk 6]
```diff
@@ -98,6 +98,16 @@ class QueryingController {
         this.sorting.loadOptions();
     }
 
+    /**
+     * Check if the data table does not need to be modified.
+     */
+    public willNotModify(): boolean {
+        return (
+            !this.sorting.modifier
+            // && !this.filtering.modifier
```
Comment (sebastianbochan) [inline ts/DataGrid/Querying/QueryingController.ts:107]: To remove

File: ts/DataGrid/Table/Cell.ts [hunk 7]
```diff
@@ -119,6 +125,47 @@ abstract class Cell {
...
+        if (this.customClassName) {
+            element.classList.remove(this.customClassName);
+        }
+
+        if (!template) {
+            delete this.customClassName;
+            return;
+        }
+
+        const newClassName = this.format(template).replace(/\s+/g, '-');
```
Comment (sebastianbochan) [inline ts/DataGrid/Table/Cell.ts:159]: `template.replace` is enough.

**Review by sebastianbochan — CHANGES_REQUESTED** [review]

File: ts/DataGrid/Querying/QueryingController.ts (hunk 6 above)
Comment (ddragula) [reply ts/DataGrid/Querying/QueryingController.ts:107]: I left it in the querying templates for filtering to make it easier to add it later once `FilterModifier` is implemented.

**Review by stitot — APPROVED** [review]
Nice!

**Review by sebastianbochan — APPROVED** [review]

**Pull request merged**